### PR TITLE
Add AsyncDDOGTracer

### LINKED ISSUE #1236: Update DDOGTracer to work with AsyncDriver
**Is your feature request related to a problem? Please describe.**
The current DDOGTracer with AsyncDriver creates a span for each Hamilton function/node when it creates a coroutine, but doesn't track when the function actually runs. It would be great to have it track timing of the actual function body execution for more accurate metrics. 

Also when using Hamilton in a FastAPI app, the FastAPI middleware tracer is not correlating the Hamilton trace as a child, i.e. 2 separate traces appear in DataDog. Ideally the Hamilton trace uses the current FastAPI request trace as its parent. 

**Describe the solution you'd like**

**Describe alternatives you've considered**
Open to using the OTel adapter instead of DDog if it's any easier to implement these changes. 

**Additional context**
Slack thread: https://hamilton-opensource.slack.com/archives/C03M33QB4M8/p1731597046007369

### PR DESCRIPTION
**Pull request by rwhitten577**
Adds `AsyncDDOGTracer`, which has all the same functionality of `DDOGTracer` but inherits from the base async lifecycle classes to make it work with the AsyncDriver. Fixes #1236. 

## Changes
* Add `AsyncDDOGTracer`
* Refactor `DDOGTracer` functionality into private `_DDOGTracerImpl`, which allows sharing logic between the sync & async versions
* Added check for current trace context so Hamilton DAG run gets properly nested under any open parent spans. This fixed the issue I had where my FastAPI route and Hamilton run were 2 independent traces in DDOG. It will be a no-op if there is no current trace. 
* Update docs to add `AsyncDDOGTracer` reference

## How I tested this
- Tested 1.83.3 to verify behavior of all nodes showing executed quickly at the beginning:
hamilton_ddog_sync_prev
 
- Tested new implementation of standard DDOGTracer to ensure it works as expected (notice Hamilton is now nested under the FastAPI parent span):
hamilton_ddog_sync_new

- Tested new async implementation: 
hamilton_ddog_async_new

## Checklist

- [x] PR has an informative and human-readable title (this will be pulled into the release notes)
- [x] Changes are limited to a single goal (no scope creep)
- [x] Code passed the pre-commit check & code is left cleaner/nicer than when first encountered.
- [x] Any _change_ in functionality is tested
- [x] New functions are documented (with a description, list of inputs, and expected output)
- [x] Placeholder code is flagged / future TODOs are captured in comments
- [x] Project documentation has been updated if adding/changing functionality.

### CODE REVIEW

**Comment by skrawcz** [comment]
Exciting! Let us know when to review (should we wait for it for this week's release?). Otherwise we'll rely on you for testing that it works for real ;)

**Comment by rwhitten577** [comment]
> Exciting! Let us know when to review (should we wait for it for this week's release?). Otherwise we'll rely on you for testing that it works for real ;)

Thanks @skrawcz, ready to review. Tested out both sync and async versions, and added some screenshots here in the description

File: hamilton/plugins/h_ddog.py [hunk 1]
```diff
@@ -130,7 +106,10 @@ def run_before_graph_execution(self, *, run_id: str, **future_kwargs: Any):
         :param run_id: ID of the run
         :param future_kwargs: reserved for future keyword arguments/backwards compatibility.
         """
-        span = tracer.start_span(name=self.root_name, activate=True, service=self.service)
+        current_context = tracer.current_trace_context()
```
Comment (rwhitten577) [inline hamilton/plugins/h_ddog.py:109]: This returns `None` if there's no active trace, so will be a safe no-op if there isn't one. `child_of` in `start_span` is `None` by default, so we're just assigning the parent if it exists.

File: hamilton/plugins/h_ddog.py [hunk 2]
```diff
@@ -267,3 +246,248 @@ def run_after_task_execution(
...
+        self._impl.run_after_node_execution(
+            node_name=node_name, error=error, task_id=task_id, run_id=run_id, **future_kwargs
+        )
+
+    def run_after_graph_execution(
+        self, *, error: Optional[Exception], run_id: str, **future_kwargs: Any
+    ):
+        """Runs after graph execution. Garbage collects + finishes the root span.
+
+        :param error: Error the graph raised when running, if any
```
[the remark this answers is not in the document]
Comment (elijahbenizzy) [reply hamilton/plugins/h_ddog.py:346]: Yes, this is specifically *not* supported (right now). I think this is good for now -- add a note about it.

**Review by elijahbenizzy — APPROVED** [review]
Looks really clean! Nice work. Only comments are about comments int he code -- a little bit of design choice can make it a bit easier to maintain.

Otherwise if you add some comments and push another one I'll merge + release shortly.

File: hamilton/plugins/h_ddog.py (hunk 1 above)
Comment (rwhitten577) [reply hamilton/plugins/h_ddog.py:119]: Good idea — added!

File: hamilton/plugins/h_ddog.py [hunk 3]
```diff
@@ -16,35 +19,8 @@
...
-    Monitoring ETLs is not a typical datadog case (you can't see relationships between nodes/tasks or data summaries),
-    but it is easy enough to work with and gives some basic information.
-
-    This tracer bypasses context management so we can more accurately track relationships between nodes/tags. Also, we plan to
-    get this working with OpenTelemetry, and use that for datadog integration.
-
-    To use this, you'll want to run `pip install sf-hamilton[ddog]` (or `pip install "sf-hamilton[ddog]"` if using zsh)
-    """
-
+class _DDOGTracerImpl:
```
[the remark this answers is not in the document]
Comment (rwhitten577) [reply hamilton/plugins/h_ddog.py:22]: Thanks, added!

**Review by elijahbenizzy — APPROVED** [review]

**Pull request merged**